diff -up dovecotcore/dovecot-init.service.initbysystemd dovecotcore/dovecot-init.service
--- dovecotcore/dovecot-init.service.initbysystemd	2024-12-03 23:26:27.694327778 +0100
+++ dovecotcore/dovecot-init.service	2024-12-03 23:26:27.694327778 +0100
@@ -0,0 +1,13 @@
+[Unit]
+Description=One-time Dovecot init service
+ConditionPathExists=|!/etc/pki/dovecot/certs/dovecot.pem
+
+[Service]
+Type=oneshot
+RemainAfterExit=no
+ExecStart=/bin/sh -c '\
+if [ ! -f /etc/pki/dovecot/certs/dovecot.pem ]; \
+then\
+  SSLDIR=/etc/pki/dovecot/ OPENSSLCONFIG=/etc/pki/dovecot/dovecot-openssl.cnf /usr/libexec/dovecot/mkcert.sh /dev/null 2>&1;\
+fi'
+
diff -up dovecotcore/dovecot.service.in.initbysystemd dovecotcore/dovecot.service.in
--- dovecotcore/dovecot.service.in.initbysystemd	2024-12-03 23:26:27.695327787 +0100
+++ dovecotcore/dovecot.service.in	2024-12-03 23:27:24.466844738 +0100
@@ -11,7 +11,8 @@
 Description=Dovecot IMAP/POP3 email server
 Documentation=man:dovecot(1)
 Documentation=https://doc.dovecot.org/
-After=local-fs.target network-online.target remote-fs.target time-sync.target
+After=local-fs.target network-online.target remote-fs.target time-sync.target dovecot-init.service
+Requires=dovecot-init.service
 Wants=network-online.target
 
 [Service]
diff -up dovecotcore/Makefile.am.initbysystemd dovecotcore/Makefile.am
--- dovecotcore/Makefile.am.initbysystemd	2024-10-15 14:46:51.987002649 +0200
+++ dovecotcore/Makefile.am	2024-12-03 23:26:27.696327797 +0100
@@ -19,6 +19,7 @@ EXTRA_DIST = \
 	run-test-valgrind.supp \
 	dovecot.service.in \
 	dovecot.socket \
+	dovecot-init.service \
 	version \
 	$(conf_DATA)
 
@@ -62,7 +63,8 @@ dovecot-config: dovecot-config.in Makefi
 if WANT_SYSTEMD
 systemdsystemunit_DATA = \
         dovecot.socket \
-        dovecot.service
+        dovecot.service \
+        dovecot-init.service
 endif
 
 install-exec-hook:
